mlflow/utils/checkpoint_utils.py

```diff
@@ -0,0 +1,182 @@
+import os
+import shutil
+import logging
+import mlflow
+from mlflow.exceptions import MlflowException
+from mlflow.utils.file_utils import create_tmp_dir
+
+from mlflow.utils.mlflow_tags import LATEST_CHECKPOINT_ARTIFACT_TAG_KEY
+from mlflow.utils.autologging_utils import (
+    ExceptionSafeAbstractClass,
+)
+
+
+_logger = logging.getLogger(__name__)
+
+
+class MlflowModelCheckpointCallbackBase(metaclass=ExceptionSafeAbstractClass):
+
+    def __init__(
+        self,
+        client,
+        run_id,
+        monitor,
+        mode,
+        save_best_only,
+        save_weights_only,
+        save_freq,
+    ):
+        """
+        Args:
+            client: An instance of `MlflowClient`.
+            run_id: The id of the MLflow run which you want to log checkpoints to.
+            monitor: In automatic model checkpointing, the metric name to monitor if
+                you set `model_checkpoint_save_best_only` to True.
+            save_best_only: If True, automatic model checkpointing only saves when
+                the model is considered the "best" model according to the quantity
+                monitored and previous checkpoint model is overwritten.
+            mode: one of {"min", "max"}. In automatic model checkpointing,
+                if save_best_only=True, the decision to overwrite the current save file is made
+                based on either the maximization or the minimization of the monitored quantity.
+            save_weights_only: In automatic model checkpointing, if True, then
+                only the model’s weights will be saved. Otherwise, the optimizer states,
+                lr-scheduler states, etc are added in the checkpoint too.
+            save_freq: `"epoch"` or integer. When using `"epoch"`, the callback
+                saves the model after each epoch. When using integer, the callback
+                saves the model at end of this many batches. Note that if the saving isn't
+                aligned to epochs, the monitored metric may potentially be less reliable (it
+                could reflect as little as 1 batch, since the metrics get reset
+                every epoch). Defaults to `"epoch"`.
+        """
+        self.client = client
```

Comment:
I prefer to use client API, it can set tracking uri explicitly (we need to set tracking uri here, for distributed training case). Any reason to use fluent API ?

Reply:
I think fluent api is our standard user interface, so it is more robust, also we can shorten the arglist. For distributed training scenario, is `mlflow.set_tracking_uri()` enough?

But anyway, I don't think this is a blocking issue, just want to open the discussion to align the API design.

Reply:
oh, we can remove `client` arguements from `MlflowModelCheckpointCallback` constructor. 

btw, MlflowClient apis should be also stable and robust, it is widely used, and it can log into run of specific run_id, which is more easier to use in internal code.